nemo/collections/tts/models/spectrogram_enhancer.py

```diff
@@ -0,0 +1,382 @@
+from nemo.collections.tts.modules.spectrogram_enhancer import Discriminator, Generator, mask
+from nemo.core import Exportable, ModelPT
+from nemo.utils import logging
+
+
+def type_as_recursive(e, source: Tensor):
+    if isinstance(e, (list, tuple)):
+        return [type_as_recursive(elem, source) for elem in e]
+    elif isinstance(e, dict):
+        return {key: type_as_recursive(value, source) for key, value in e.items()}
+    elif isinstance(e, Tensor):
+        return e.type_as(source)
+    else:
+        return e
+
+
+def gradient_penalty(images, output, weight=10):
+    batch_size = images.shape[0]
+    gradients = torch_grad(
+        outputs=output,
+        inputs=images,
+        grad_outputs=torch.ones(output.size(), device=images.device),
+        create_graph=True,
+        retain_graph=True,
+        only_inputs=True,
+    )[0]
+
+    gradients = gradients.reshape(batch_size, -1)
+    return weight * ((gradients.norm(2, dim=1) - 1) ** 2).mean()
+
+
+def calc_pl_lengths(styles, images):
+    device = images.device
+    num_pixels = images.shape[2] * images.shape[3]
+    pl_noise = torch.randn(images.shape, device=device) / math.sqrt(num_pixels)
+    outputs = (images * pl_noise).sum()
+
+    pl_grads = torch_grad(
+        outputs=outputs,
+        inputs=styles,
+        grad_outputs=torch.ones(outputs.shape, device=device),
+        create_graph=True,
+        retain_graph=True,
+        only_inputs=True,
+    )[0]
+
+    return (pl_grads ** 2).sum(dim=2).mean(dim=1).sqrt()
+
+
+def gen_hinge_loss(fake):
+    return fake.mean()
+
+
+def hinge_loss(real, fake):
+    return (F.relu(1 + real) + F.relu(1 - fake)).mean()
+
+
+def consistency_loss(condition, output, lengths):
+    *_, w, h = condition.shape
+    w, h = w // 4, h
+
+    condition = torch.nn.functional.interpolate(condition, size=(w, h), mode="bilinear", antialias=True)
+    output = torch.nn.functional.interpolate(output, size=(w, h), mode="bilinear", antialias=True)
+
+    dist = (condition - output).abs()
+    dist = mask(dist, lengths)
+    return (dist / rearrange(lengths, "b 1 -> b 1 1 1")).sum(dim=-1).mean()
+
+
+class SpectrogramEnhancerModel(ModelPT, Exportable):
+    """GAN-based model to add details to blurry spectrograms from TTS models like Tacotron or FastPitch."""
+
+    def __init__(self, cfg: DictConfig, trainer: Trainer = None) -> None:
+        self.spectrogram_model = None
+        super().__init__(cfg=cfg, trainer=trainer)
+
+        self.init_spectrogram_model()
+
+        self.G = Generator(
+            n_bands=cfg["n_bands"],
+            latent_dim=cfg["latent_dim"],
+            network_capacity=cfg["network_capacity"],
+            style_depth=cfg["style_depth"],
+            channels=1,
+            fmap_max=cfg["fmap_max"],
+        )
+        self.D = Discriminator(
+            n_bands=cfg["n_bands"], network_capacity=cfg["network_capacity"], channels=1, fmap_max=cfg["fmap_max"],
+        )
+
+    def init_spectrogram_model(self):
```

Comment:
Do we really need a `spectrogram_model`, can this model not get spectrograms directly from the dataloader? Look at Vocoder dataloader. Or if you want to use TTSDataset class then I think you can use "log_mel". The spectrograms can be precomputed, this should remove dependency on any particular spectrogram generator.

Reply:
This is a major part of the training process, we can't do without a spectrogram model
![sg2-enhancer-training-150](https://user-images.githubusercontent.com/8864149/207025284-efd9cf2f-c2c1-421f-b2d6-05eec84d06c7.png)

Reply:
Well, essentially what you are doing in this model is mel to mel. The goal is to not have any dependency on the spectrogram generator. Here FastPitch is frozen and is not trained as part of the spectrogram enhancer.

In HiFiGAN training also the input is mel, which are precomputed, which can be done here as well. 
Before training, do inference on FastPitch or Any spectrogram generator model to precompute the mels. Now, these mels can be an input to the spectrogram enhancer. In your PR as well it is part of the `prepare_batch` method, which means it is not part of training but used in preparing data, which in turn means this can be precomputed. See the https://github.com/NVIDIA/NeMo/blob/main/nemo/collections/tts/torch/data.py#L995 part of the VocoderDataset class that loads precomputed spectrograms.

During training your batch can look like this: `(precomputed_spec, groundtruth audio, len, ...)`
You can use `LogMel` ttsdatatype to load your mels and incorporate the padding logic - https://github.com/NVIDIA/NeMo/blob/main/nemo/collections/tts/torch/data.py#L514

Reply:
No. Precomputing spectrograms is very inconvenient (and also requires lots of space). I speak from experience.

Current implementation is the most clean and convenient way of doing this from user point of view.

Reply:
The main goal was to completely repeat data setup from FastPitch training: to use the same texts, audio and ground-truth precomputed features like pitch (and I guess energy, with new revisions). This gets us would get natural-sythesize) spectrogram pairs and minimizes chances of the GAN learning to fix pitch or other features. 
I figured the most easy and natural way to implement this is just to run FastPitch on its underlying TTSDataset, which is the intended way to train this model.

The initial precomputing script I used just initialized the TTSDataset and iterated through it, saving results. Same thing happening here, but inside PTL.

I haven't measured overhead, but I also did not notice any significant slowdowns while doing experiments for our paper
compared to precomputed spectrograms. Slower training is solved by waiting a bit more, but lack of storage is solved only by buying more storage.
Same thing with GPU memory though... I'll look at memory usage and get back to you on this

Note that `generate_mels.py` mirrors some of TTSDataset internals but does not use it directly. Handling of sup data mangled names is a pain, so it would be a full rewrite or just a separate script.

Reply:
Also, one very important aspect of spectrogram enhancer is that it should be independent of spectrogram generator. Why do we want to tie to any particular spectrogram generator (FastPitch)? If we want it to work with FastPitch only then we should make it a feature of FastPitch class. 

Having precomputed mels will enable us to use it for any spectrogram generator model. I understand that the experiments were done using FastPitch as spec generator, but when we release it on NeMo we should make it generic (so that it can be used with any spec gen model) as much as possible in my opinion.

Reply:
> Note that generate_mels.py mirrors some of TTSDataset internals but does not use it directly. Handling of sup data mangled names is a pain, so it would be a full rewrite to TTSDataset (questionable) or just a separate script.

`generate_mels.py` is just doing inference to generate the mels. You don't really need any other supplementary data other than `attn_prior` and `speaker` both of which are supplied. Pitch, energy, duration etc are predicted by the model.

Reply:
> Why do we want to tie to any particular spectrogram generator (FastPitch)? 

We really do not, it's just an example we're discussing. To use any other spectrogram generator just implement the boilerplate for converting a collated batch from TTSDataset to batch of paired real-fake spectrograms and it's good to go. Unfortunately they have different inputs (and interfaces) and thus need to be supported separately.

> generate_mels.py is just doing inference to generate the mels. You don't really need any other supplementary data other than attn_prior and speaker both of which are supplied. Pitch, energy, duration etc are predicted by the model.

Yes, and that's why it is not suitable for training this model. The more ground-truth features (pitch, energy), the better; durations also need to match. We need resynthesis using a TTS model.

Reply:
There should not be any functional difference from the approaches. I guess generate_mels.py is not using ground truth information, but it should have it as a configuration. When doing GAN based fine-tuning with the spectrogram (whether through StyleGan or indirectly HiFi-Gan fine-tune) it is not 100% clear to me which would be better. Using ground truth feature input is definitely safer but can limit the potential quality improvements.

So precomputing and making generate_mels.py more generic is one possible solution. It sounds like the other solution you are proposing is to turn "prepare_batch(self, batch)" into an abstract method or separate class that someone can implement with model-specific logic.